### ISSUE TITLE
Learning rate scheduler in C++ API

### ISSUE THREAD

**Issue by jamesbut** [pytorch/pytorch#50577]
Hey!

Am I right in thinking that the learning rate scheduler is not available in the C++ API?

If not, are there any plans to implement it? Is there anything I can do to help with this?

Thanks 
James

cc @yf225 @glaringlee @vincentqb

**Comment by glaringlee** [comment]
@jamesbut 
I'll take a look at this and get u back soon!

**Comment by glaringlee** [comment]
@jamesbut Spent sometime looking into the code.

The current status is that, we have multiple optimizers implemented in c++ based on previous user requests. But we don't have any scheduler implementation at this point. 

We currently don't have plan to implement this in the first half of this year,  but I looked through our current repo, the Optimizer class is already there, we need a Scheduler class as base and expend it to the specific class you need. We can write a wrapper class for all kinds of Optimizers to have counter. 

The tricky part is that , the parameter group currently is a vector, but lr_scheduler needs a list of initial base learning rate from the input optimizer's parameter group which need the parameter group be a dict, one way to solve this is to change the Optimizer adding a learning rate list (or similar class, etc).

If you are interested, you can start on writing a Scheduler base class.

**Comment by jamesbut** [comment]
Yes I agree, we need a Scheduler base class and then extensions. 

I ended up implementing my own version at: [https://github.com/jamesbut/NeuroEvo/tree/master/include/util/torch/schedulers](https://github.com/jamesbut/NeuroEvo/tree/master/include/util/torch/schedulers)

But the problem here is that OptimizerOptions does not have learning rate as a member variable - so I have to cast all the different types of options and then get references to the learning rates of all the parameter groups.

We don't need to do this though - we could move the learning rate member variable into OptimizerOptions (all optimiser options so far use learning rates) and then in the Scheduler implementation one can take a reference to the Optimiser and iterate over all the group params OptimizerOptions and set the learning rate; this is what I have done in my code. Of course we can also add the option to give a learning rate _list_ and have different learning rates for different parameter groups but I imagine the main use case it to change all param groups to the same learning rate.

I would be very happy to implement this, do you want me to help? Should I submit a pull request?

James

**Comment by glaringlee** [comment]
@jamesbut 
Yep, we can change OptimizerOptions for sure. 
You are very welcome to submit a pull request to us! 
Please hold for this weekend, I will take a look at your current impl and comment a bit, and you can submit a PR to us next week. 
Let me know if this is good for you?

**Comment by jamesbut** [comment]
That sounds good :).

FYI - the part of my implementation that casts the individual options and gets references to their learning rates (i.e. the function ascertain_param_group_lr_refs()) will not be needed in the pull request once learning rate is part of the OptimiserOptions base class. 

Also, in my implementation I have made it such that you can only set all param groups to the same learning rate; in the pull request, I will write it such that you can provide a vector where each element will correspond to a learning rate for the respective param group.

**Comment by ssbotelh** [comment]
Hi @jamesbut @glaringlee, thanks for working on this feature. I'm very interested in using LR schedulers in libtorch, and was about to start implementing my own solution. Do you have any idea when this work will be completed, and whether it will intercept the next libtorch release? I just need to decide if I can afford to wait, or if I need to get going with my implementation. Thanks!

**Comment by jamesbut** [comment]
Hey @ssbotelh,

For now you can use the [code](https://github.com/jamesbut/NeuroEvo/tree/master/include/util/torch/schedulers) that I wrote. I have only implemented ReduceLROnPlateau and StepLR for now but you can extend the base class and add your own schedulers too.

Also, this will only work for Adam right now but you can easily extend this to any of the other optimisers by filling in the appropriate casts in ascertain_param_group_lr_refs() in learning_rate_scheduler.cpp

**Comment by ssbotelh** [comment]
Thanks, @jamesbut. Yes, it would be pretty simple to extend your code. However, depending on the timeline of your work, it would still make more sense to wait for the next release. Could you please let me know what your timeline is for finishing the work, and what Pytorch release are you planning on intercepting? Thanks again!

**Comment by jamesbut** [comment]
I am not sure on the timeline quite yet but I will let you know when I have a better idea.

**Comment by glaringlee** [comment]
@jamesbut 
Sry about the delay, I was sick during the weekend and probably need rest today as well. I will go over this code no later then tomorrow and update here. Then we can start an official PR to the pytorch repo.

@jamesbut @ssbotelh 
pytorch 1.8 deadline is already passed. I am targeting to include this into pytorch 1.9. But welcome to use it immediately once it is landed into pytorch master repo.

**Comment by glaringlee** [comment]
@jamesbut 
Hi, I went through the lr scheduler implementation. Here are some comments.

1. For base class. One specific problem is `_step_count`.  C++ is not that flexible as python. We have no way to hookup a setp_count to a class object of c++.  One way to support this is to add step_count in optimizer base class as well, but need to change all the optimizers' step function to increase this count (not many though). 
Given this is to prevent a behavior change in pytorch 1.1 which is long time ago, I am open on whether to support this. Let me know, I can help on changing all existing c++ optimizers' step function to increase the `step_count`.

2. Please follow the naming convention of our python as much as possible. For eg. LearningRateScheduler -> LRScheduler ('_' is not needed)
Here is the major python optimizers' implementations.  You can find the names there.
https://github.com/pytorch/pytorch/blob/e6779d4357ae94cc9f9fedb83a87eb6126016769/torch/optim/lr_scheduler.py

3. For the scheduler that doesn't have a python version, eg, PerEpochLR, we have to add both python and c++ version since the we want the c++ code always be parity with python. I think most of the schedulers are already included in pytorch. PerEpochLR might be the only one that not included. But I think this one is a bit hacky, you might just keep it for you own usage?

4. Tests. Each scheduler need a test. We currently do gtest here,
https://github.com/pytorch/pytorch/tree/e6779d4357ae94cc9f9fedb83a87eb6126016769/test/cpp/api
You can take a look at optim.cpp (or any cpp files) and the readme. It is pretty standard.

Please take a look at above, feel free to change OptimizerOptions.  

And welcome to go ahead submitting a PR to pytorch repo!

**Comment by jamesbut** [comment]
@glaringlee 
Thank you for looking at my code.

1. I do not understand how this is a problem? Can we not have step count in the LRScheduler bass class like in python?

3. Sure the PerEpochLR was a hacky implementation that I was using for my own research - I will leave this out.

I will follow the naming conventions and write some tests :).

**Comment by glaringlee** [comment]
@jamesbut 
Hi, in python LRScheduler do have step_count, see this line.
https://github.com/pytorch/pytorch/blob/e6779d4357ae94cc9f9fedb83a87eb6126016769/torch/optim/lr_scheduler.py#L74
And it hook up a step_count to each passed in optimizer instance. see here:
https://github.com/pytorch/pytorch/blob/e6779d4357ae94cc9f9fedb83a87eb6126016769/torch/optim/lr_scheduler.py#L72
This is doable in python (though hacky.....), but no C++.

This is to make sure optimizer's step() is called before LRScheduler since we had a major change (Backward compatibility break) back to pytorch 1.1, this is to ensure user adopt the new behavior correctly. 

To support his. One way is to change all optimizers adding set_count, the other way is to write a small wrapper class that wraps the optimizer (use optimizer base ptr) and add set_count to the wrap. I think the 1st way might be more generic.

I looked through the python code, the only place we need that step_count is here.
https://github.com/pytorch/pytorch/blob/e6779d4357ae94cc9f9fedb83a87eb6126016769/torch/optim/swa_utils.py#L259

I am open on whether to add step_count to Optimizer's, but let's at least add step_count to the LRScheduler.

**Comment by jamesbut** [comment]
@glaringlee 

Thank you for the reply :). I am working on this now, I will let you know if I have any more questions.

Best
James

**Comment by jamesbut** [comment]
@glaringlee 

I have come up against a few problems. 

If we move TORCH_ARG(double, lr) into the OptimizerOptions base class then the following behaviour/use case doesn't work:

AdamOptions().lr(0.999).amsgrad(true);

Because of the definition of TORCH_ARG(), when one calls name(const T& new_##name)->decltype(*this), *this is returned which in the above case is OptimiserOptions not AdamOptions. 

This is a problem because a number of tests in serialize.cpp do not compile because they rely on lr(0.999) returning the sub class not the base class. I also guess we want to keep the AdamOptions().lr() behaviour because it is the same as the Python version.

There seems to be a few options:

1. Keep TORCH_ARG(double, lr) in all the sub classes like they are now but put an abstract virtual function get_lr_ref() in OptimizerOptions, then anybody writing a subclass will have to explicitly implement this method and just return a reference to the learning rate. Bear in mind we need access to learning rate from OptimiserOptions in order to modify it using the LRScheduler.

2. We can put TORCH_ARG(double, lr) in OptimiserOptions and have it such that the sub classes can only set the learning rate via their constructors, which is an option now, but by using this option one will not be able to call AdamOptions.lr(0.999) for example.

I can't yet think of a way to do this without breaking some behaviour, what do you think?

**Comment by jamesbut** [comment]
If I implement it like I highlighted in option 1. above then we have 3 options:

1. Implement the getters and setters as abstract virtual functions so that one cannot compile the code without overriding.
2. Implement them as just a virtual functions with a runtime written warning that they should override the function.
3. Implement them as just a virtual function with a TORCH_CHECK such that the code exits at this point.

Options 2 and 3 make it such that it is backward compatible (in that it will still compile), and calls such as AdamOptions.lr(0.999) still work.

**Comment by glaringlee** [comment]
@jamesbut 
I think BC compatibility is important. So we should go with option 1 and then choose between option 2 and 3. 
I will do some research today see whether we have better options, will get back to you later today or tomorrow.
Thanks for contributing this!!!!

**Comment by glaringlee** [comment]
@jamesbut Sry about the lateness, was woking on some high priority benchmarking thing previously.
Let's do option 1. Adding get_lr_ref() with TORCH_CHECK (to prevent wrong usage).

**Comment by jamesbut** [comment]
@glaringlee 

No worries :) - Yes sure, I have written the main code, I am just getting some tests together so I should be able to submit the pull request soon
